Framework/DataHandling/src/LoadNGEM.cpp

```diff
@@ -0,0 +1,474 @@
+ * @param frameEventCounts A vector of the number of events in each good frame.
+ * @param histograms The main set of histograms for the data so far.
+ * @param histogramsInFrame The set of histograms for the current frame.
+ * @param totalFilePaths The total number of file paths.
+ * @param fileCount The number of file paths processed.
+ */
+void LoadNGEM::loadSingleFile(
+    const std::vector<std::string> &filePath, int &eventCountInFrame,
+    int &maxToF, int &minToF, int &rawFrames, int &goodFrames,
+    const int &minEventsReq, const int &maxEventsReq,
+    MantidVec &frameEventCounts,
+    std::vector<DataObjects::EventList> &histograms,
+    std::vector<DataObjects::EventList> &histogramsInFrame,
+    const size_t &totalFilePaths, int &fileCount) {
+  // Create file reader
+  if (filePath.size() > 1) {
+    throw std::runtime_error("Invalid filename parameter.");
+  }
+  std::ifstream file(filePath[0].c_str(), std::ifstream::binary);
+  if (!file.is_open()) {
+    throw std::runtime_error("File could not be found.");
+  }
+  std::array<char, 16> buffer;
+
+  const size_t totalNumEvents = verifyFileSize(file) / 16;
+  size_t numProcessedEvents = 0;
+
+  while (!file.eof()) {
+    // Load an event into the variable.
+    file.read(buffer.data(), 16);
+    auto eventBigEndian = reinterpret_cast<EventUnion *>(buffer.data());
+
+    // Correct for the big endian format.
+    EventUnion event;
+    correctForBigEndian(eventBigEndian, event);
+
+    if (event.coincidence.check()) { // Check for coincidence event.
+      ++eventCountInFrame;
+      uint64_t pixel = event.coincidence.getPixel();
+      int tof =
+          event.coincidence.timeOfFlight / 1000; // Convert to microseconds (us)
+
+      if (tof > maxToF) {
+        maxToF = tof;
+      } else if (tof < minToF) {
+        minToF = tof;
+      }
+      histogramsInFrame[pixel].addEventQuickly(
+          Mantid::Types::Event::TofEvent(tof));
+
+    } else if (event.tZero.check()) { // Check for T0 event.
+      addFrameToOutputWorkspace(rawFrames, goodFrames, eventCountInFrame,
+                                minEventsReq, maxEventsReq, frameEventCounts,
+                                histograms, histogramsInFrame);
+
+      if (reportProgressAndCheckCancel(numProcessedEvents, eventCountInFrame,
+                                       totalNumEvents, totalFilePaths,
+                                       fileCount)) {
+        return;
+      }
+    } else { // Catch all other events and notify.
+      g_log.warning() << "Unexpected event type loaded.\n";
+    }
+  }
+  g_log.information() << "Finished loading a file.\n";
+  ++fileCount;
+}
+
+/**
+ * @brief Correct a big endian event to be compatible with x64 and x86
+ * architecture.
+ *
+ * @param bigEndian The big endian formatted event.
+ * @param smallEndian The resulting small endian formatted event.
+ */
+void LoadNGEM::correctForBigEndian(EventUnion *&bigEndian,
```

Comment:
There are a few functions like this that do not depend on any state in the class and could easily be free functions.

Effective C++ has a good explanation of why this should be preferred.